File: ASAttributedLabelNode.m
//
//  ASAttributedLabelNode.h
//
//  Created by Peter Easdown on 27/11/16.
//  Inspired by Alex Studnicka
//

#import "ASAttributedLabelNode.h"

@implementation ASAttributedLabelNode

- (id) initWithSize:(CGSize)size {
    self = [super initWithColor:[UIColor clearColor] size:size];
    
    return self;
}

- (void) setAttributedText:(NSAttributedString *)attributedText {
    _attributedText = attributedText;
    
    [self draw];
}

- (void) draw {
    NSAttributedString *attrStr = self.attributedText;
    
    if (attrStr == nil) {
        self.texture = nil;
        
        return;
    }
    
    float scaleFactor = [UIScreen mainScreen].scale;
    CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceRGB();
    CGImageAlphaInfo bitmapInfo = kCGImageAlphaPremultipliedLast;
    CGContextRef context = CGBitmapContextCreate(nil, self.size.width * scaleFactor, self.size.height * scaleFactor, 8, (self.size.width * scaleFactor) * 4, colorSpace, bitmapInfo);
    
    if (context == nil) {
        return;
    }

    CGContextScaleCTM(context, scaleFactor, scaleFactor);
    CGContextConcatCTM(context, CGAffineTransformMake(1, 0, 0, -1, 0, self.size.height));
    UIGraphicsPushContext(context);
    
    float strHeight = [attrStr boundingRectWithSize:self.size options:NSStringDrawingUsesLineFragmentOrigin context:nil].size.height;
    float yOffset = (self.size.height - strHeight) / 2.0;
    CGRect rect = CGRectMake(0.0, yOffset, self.size.width, strHeight);
    [attrStr drawWithRect:rect options:NSStringDrawingUsesLineFragmentOrigin context:nil];
    
    CGImageRef imageRef = CGBitmapContextCreateImage(context);
    
    if (imageRef != nil) {
        self.texture = [SKTexture textureWithCGImage:imageRef];
        CGImageRelease(imageRef);
    }
    
    UIGraphicsPopContext();
}

@end

File: ASAttributedLabelNode.swift
//
//  ASAttributedLabelNode.swift
//
//  Created by Alex Studnicka on 15/08/14.
//  Copyright © 2016 Alex Studnicka. MIT License.
//

import UIKit
import SpriteKit

class ASAttributedLabelNode: SKSpriteNode {
	
	required init?(coder aDecoder: NSCoder) {
		super.init(coder: aDecoder)
	}
	
	init(size: CGSize) {
		super.init(texture: nil, color: UIColor.clear, size: size)
	}
	
	var attributedString: NSAttributedString! {
		didSet {
			draw()
		}
	}
	
	func draw() {
		guard let attrStr = attributedString else {
			texture = nil
			return
		}
		
		let scaleFactor = UIScreen.main.scale
		let colorSpace = CGColorSpaceCreateDeviceRGB()
		let bitmapInfo = CGImageAlphaInfo.premultipliedLast.rawValue
		guard let context = CGContext(data: nil, width: Int(size.width * scaleFactor), height: Int(size.height * scaleFactor), bitsPerComponent: 8, bytesPerRow: Int(size.width * scaleFactor) * 4, space: colorSpace, bitmapInfo: bitmapInfo) else {
			return
		}

		context.scaleBy(x: scaleFactor, y: scaleFactor)
		context.concatenate(CGAffineTransform(a: 1, b: 0, c: 0, d: -1, tx: 0, ty: size.height))
		UIGraphicsPushContext(context)
		
		let strHeight = attrStr.boundingRect(with: size, options: .usesLineFragmentOrigin, context: nil).height
		let yOffset = (size.height - strHeight) / 2.0
		attrStr.draw(with: CGRect(x: 0, y: yOffset, width: size.width, height: strHeight), options: .usesLineFragmentOrigin, context: nil)
		
		if let imageRef = context.makeImage() {
			texture = SKTexture(cgImage: imageRef)
		} else {
			texture = nil
		}
		
		UIGraphicsPopContext()
	}
	
}
